src/set.cr

```diff
@@ -152,6 +152,30 @@ struct Set(T)
     set
   end
 
+  # Difference: returns a new set containing elements in this set that are not
+  # present in the other.
+  #
+  #     Set.new([1,2,3,4,5]) - Set.new([2,4])               #=> Set{1, 3, 5}
+  #     Set.new(['a','b','b','z']) - Set.new(['a','b','c']) #=> Set{'z'}
+  def -(other : Set)
+    set = Set(T).new
+    each do |value|
+      set.add value unless other.includes?(value)
+    end
+    set
+  end
```

Comment:
Regarding the exact type, the same is true by restricting upon `Set` instead of `self` already though and yes I'd say it's preferred in this case to allow sets with distinct types.

My main motivation is that something like `set - some_array` is much more succinct compared to `set - Set.new(some_array)`. I guess the interesting part would be where the overhead of allocating a new set outweighs the penalty from `includes?` being O(n)